cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java

```diff
@@ -642,6 +659,42 @@ protected JSONObject normalizeRemotePrincipal(final OpenScienceFrameworkCredenti
         return XML.toJSONObject(writer.getBuffer().toString());
     }
 
+    /**
+     * Retrieve the department value from the raw department string.
+     *
+     * @param departmentRaw the raw department string
+     * @param eduPerson whether the department attribute uses eduPerson schema
+     * @return the department value
+     */
+    private String retrieveDepartment(final String departmentRaw, final boolean eduPerson) {
+
+        // Return the raw value as it is if institutions do not use eduPerson schema for the department attribute
+        if (!eduPerson) {
+            return departmentRaw;
+        }
+
+        // For institutions that use the eduPerson schema, the department must be retrieved from the raw value. Here is
+        // an example: "ou=Music Department, o=Notre Dame, dc=nd, dc=edu", whose syntax is LDAP Distinguished Names.
+        try {
+            final LdapName dn = new LdapName(departmentRaw);
+            for (int i = dn.size() - 1; i >=0; i--) {
+                final String rdn = dn.get(i);
+                if (rdn.startsWith(LDAP_DP_OU_PREFIX)) {
+                    return rdn.substring(LDAP_DP_OU_PREFIX.length()-1);
```

Comment:
A bad mistake on `substring()`. The **starting index is inclusive** but the count **starts at 1** instead of 0. Thus it should be:

```java
return rdn.substring(LDAP_DP_OU_PREFIX.length());
```